### ISSUE TITLE
React state does not use generic state type

### ISSUE THREAD

**Issue by SamHH** [DefinitelyTyped/DefinitelyTyped#27248]
- [x] I tried using the `@types/xxxx` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @johnnyreilly @bbenezech @pzavolinsky> @digiguru @ericanderson @morcerf @tkrotoff @DovydasNavickas @onigoetz @theruther4d @guilhermehubner @ferdaber @jrakotoharisoa @pascaloliv @hotell

---

Sorry if this was already posted, couldn't find it.

Example:

```typescript
interface State {
    value: number;
    tags: string[];
}

class Test extends React.Component<{}, State> {
    state = {
        value: 0,
        tags: [],
    }; // state is now { value: number; tags: never[]; } instead of State
}
```

You can also take this further; if you define `State['tags']` as `Array<string | number>` and initialise it in `state` as `['a string']`, it infers to `string[]`, which is also incorrect.

You can assert the type, but that seems needless given that we have access to the `State` generic.

**Comment by Hotell** [comment]
That's how TypesScript works :) 

Fort this particular reason this PR was introduced https://github.com/DefinitelyTyped/DefinitelyTyped/pull/26813

which explains all why's ( it also enforces you to set state via class property, not via constructor ). After huge fiasko this PR was later reverted...

**TL;DR:**

You need to set initial state within constructor to be able to apply generics declaration, as that is how TS class definition resolution works

```tsx
// Use this to get State generic applied on state
class TestViaCtor extends React.Component<{}, State> {
  constructor(props: {}) {
    super(props)

 // state is type of State from generic definition, applied via inheritance class type chain
    this.state = {
      value: 0,
      tags: [],
    }
  }
}


// GOOD
// set state via class property bud you need to explicitly annotate that array, [] infers to never[]
class TestStateViaProp extends React.Component<{}, State> {
  state = {
    value: 0,
    tags: [] as string[],
  }
}

// BETTER
// you need to anotate state property again
class TestStateViaProp extends React.Component<{}, State> {
  state: State = {
    value: 0,
    tags: [],
  }
}
```

**Comment by SamHH** [comment]
Thanks for the link!
